cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -355,25 +459,27 @@ index<T, IdxT> build(
   knn_graph.reset();
 
   RAFT_LOG_INFO("Graph optimized, creating index");
+
   // Construct an index from dataset and optimized knn graph.
-  if (construct_index_with_dataset) {
-    if (params.compression.has_value()) {
-      RAFT_EXPECTS(params.metric == cuvs::distance::DistanceType::L2Expanded,
-                   "VPQ compression is only supported with L2Expanded distance mertric");
-      index<T, IdxT> idx(res, params.metric);
-      idx.update_graph(res, raft::make_const_mdspan(cagra_graph.view()));
-      idx.update_dataset(
-        res,
-        // TODO: hardcoding codebook math to `half`, we can do runtime dispatching later
-        cuvs::neighbors::vpq_build<decltype(dataset), half, int64_t>(
-          res, *params.compression, dataset));
-
-      return idx;
-    }
+  if (params.compression.has_value()) {
+    RAFT_EXPECTS(params.metric == cuvs::distance::DistanceType::L2Expanded,
+                 "VPQ compression is only supported with L2Expanded distance mertric");
+    index<T, IdxT> idx(res, params.metric);
+    idx.update_graph(res, raft::make_const_mdspan(cagra_graph.view()));
+    idx.update_dataset(
+      res,
+      // TODO: hardcoding codebook math to `half`, we can do runtime dispatching later
+      cuvs::neighbors::vpq_build<decltype(dataset), half, int64_t>(
+        res, *params.compression, dataset));
+
+    return idx;
+  }
+  try {
     return index<T, IdxT>(res, params.metric, dataset, raft::make_const_mdspan(cagra_graph.view()));
-  } else {
-    // We just add the graph. User is expected to update dataset separately. This branch is used
-    // if user needs special control of memory allocations for the dataset.
+  } catch (std::bad_alloc& e) {
+    RAFT_LOG_DEBUG("Insufficient GPU memory to construct CAGRA index with dataset on GPU");
```

Comment:
Automatically construct index without dataset if dataset would not fit GPU memory.